src/Database/Esqueleto/PostgreSQL/WindowFunction.hs

```diff
@@ -0,0 +1,108 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeFamilies #-}
+
+module Database.Esqueleto.PostgreSQL.WindowFunction
+    ( Window, Frame, PartitionBy
+    , WindowExpr, liftExpr
+    , SqlWindow
+    , over_, rowNumber_
+    , frame_, partitionBy_, orderBy_
+    , range, rows, groups
+    , excludeCurrentRow, excludeGroup, excludeTies, excludeNoOthers
+    , between_, unboundedFollowing, unboundedPreceding, preceding, following, currentRow
+    )
+    where
+
+import Database.Esqueleto.Internal.Internal
+       ( AggregateContext
+       , NeedParens(..)
+       , SqlExpr
+       , SqlExpr_(..)
+       , Value(..)
+       , noMeta
+       , parens
+       , parensM
+       , unsafeSqlValue
+       , veryUnsafeCoerceSqlExpr
+       )
+import Database.Esqueleto.PostgreSQL.Window
+       ( Frame
+       , PartitionBy
+       , RenderWindow(..)
+       , Window
+       , between_
+       , currentRow
+       , excludeCurrentRow
+       , excludeGroup
+       , excludeNoOthers
+       , excludeTies
+       , following
+       , frame_
+       , groups
+       , orderBy_
+       , partitionBy_
+       , preceding
+       , range
+       , rows
+       , unboundedFollowing
+       , unboundedPreceding
+       )
+
+data WindowContext
+liftExpr :: SqlExpr a -> SqlExpr_ WindowContext a
+liftExpr = veryUnsafeCoerceSqlExpr
+
+-- | Helper type indicating that the 'SqlExpr_' is for a 'WindowContext'.
+type SqlWindow = SqlExpr_ WindowContext
+
+newtype WindowExpr a = WindowExpr { unWindowExpr :: SqlExpr a }
```

Comment:
When do I want to use `WindowExpr` over `SqlExpr_ WindowContext`?

Reply:
A WindowExpr is used for a window function like lag/lead/row_number

A SqlExpr_ WindowContext is a windowed result, it's existence is to prevent double windowing